src/Components/Components/src/Forms/EditContext.cs

```diff
@@ -0,0 +1,185 @@
+        /// </summary>
+        /// <param name="model">The model object for the <see cref="EditContext"/>. This object should hold the data being edited, for example as a set of properties.</param>
+        public EditContext(object model)
+        {
+            // The only reason we disallow null is because you'd almost always want one, and if you
+            // really don't, you can pass an empty object then ignore it. Ensuring it's nonnull
+            // simplifies things for all consumers of EditContext.
+            Model = model ?? throw new ArgumentNullException(nameof(model));
+        }
+
+        /// <summary>
+        /// An event that is raised when a field value changes.
+        /// </summary>
+        public event EventHandler<FieldIdentifier> OnFieldChanged;
+
+        /// <summary>
+        /// An event that is raised when validation is requested.
+        /// </summary>
+        public event EventHandler OnValidationRequested;
+
+        /// <summary>
+        /// An event that is raised when validation state has changed.
+        /// </summary>
+        public event EventHandler OnValidationStateChanged;
+
+        /// <summary>
+        /// Supplies a <see cref="FieldIdentifier"/> corresponding to a specified field name
+        /// on this <see cref="EditContext"/>'s <see cref="Model"/>.
+        /// </summary>
+        /// <param name="fieldName">The name of the editable field.</param>
+        /// <returns>A <see cref="FieldIdentifier"/> corresponding to a specified field name on this <see cref="EditContext"/>'s <see cref="Model"/>.</returns>
+        public FieldIdentifier Field(string fieldName)
+            => new FieldIdentifier(Model, fieldName);
+
+        /// <summary>
+        /// Gets the model object for this <see cref="EditContext"/>.
+        /// </summary>
+        public object Model { get; }
+
+        /// <summary>
+        /// Signals that the value for the specified field has changed.
+        /// </summary>
+        /// <param name="fieldIdentifier">Identifies the field whose value has been changed.</param>
+        public void NotifyFieldChanged(FieldIdentifier fieldIdentifier)
+        {
+            GetFieldState(fieldIdentifier, ensureExists: true).IsModified = true;
+            OnFieldChanged?.Invoke(this, fieldIdentifier);
+        }
+
+        /// <summary>
+        /// Signals that some aspect of validation state has changed.
+        /// </summary>
+        public void NotifyValidationStateChanged()
+        {
+            OnValidationStateChanged?.Invoke(this, null);
+        }
+
+        /// <summary>
+        /// Clears any modification flag that may be tracked for the specified field.
+        /// </summary>
+        /// <param name="fieldIdentifier">Identifies the field whose modification flag (if any) should be cleared.</param>
+        public void MarkAsUnmodified(FieldIdentifier fieldIdentifier)
+        {
+            if (_fieldStates.TryGetValue(fieldIdentifier, out var state))
```

Comment:
Is there a reason why we have an `if` here and `NotifyFieldChanged` doesn't? What should be the behaviour of these methods when a field can't be found?

Reply:
Yes, it's because:

 1. We want "mark a field unmodified when it's already unmodified" to be a no-op (not an error)
 2. The `EditContext` stores the field states sparsely, so there isn't necessarily any field state tracked for a given field. Especially not if it's already "unmodified", which is the default state.

> What should be the behaviour of these methods when a field can't be found?

That's not really a meaningful case. All possible field identifiers can be "found", because EditContext will track the states for any `FieldIdentifier` value you give to it. The "if not found" checks are just internal implementation details because we track the states sparsely.

I've added a comment summarizing this.

Reply:
Ok great, that's a really good reason for the current approach.